lib/pbench/test/functional/agent/cli/commands/tools/test_clear_tools.py

```diff
@@ -165,48 +199,200 @@ def test_clear_tools_test69(monkeypatch, agent_config, pbench_run, pbench_cfg):
+
+    out, err, exitcode = pytest.helpers.capture(command)
+
+    assert b"" == out
+    for tool in tools_existence:
+        tool_str = str(tool).replace(str(pbench_run), "")
+        group_check = "default" in tool_str
+        remote_check = True
+        tool_check = True
+        if groups:
+            group_check = any(
+                x in tool_str for x in [x for x in groups.split(",") if x]
+            )
+        if remotes:
+            remote_check = any(
+                x in tool_str for x in [x for x in remotes.split(",") if x]
+            )
+        if tools:
+            tool_check = any(x in tool_str for x in [x for x in tools.split(",") if x])
+
+        if tool_check and remote_check and group_check:
+            assert not tool.exists()
+            # Default group should exist even if all the remotes are gone
+            if "default" in tool_str:
+                assert tool.parent.parent.exists()
+            # if custom group exists then we need to make sure at least one
+            # of the remote is not empty
+            elif tool.parent.parent.exists():
+                for remote_dir in tool.parent.parent.iterdir():
+                    assert any(
+                        remote_dir.iterdir()
+                    ), f'Found an empty "remote" directory, "{remote_dir!s}", which should have been deleted.'
+        else:
+            assert tool.exists()
+
+
+@pytest.mark.parametrize(
+    "groups", ["", "default,wrong_group", "group1,wrong_group"],
+)
+@pytest.mark.parametrize(
+    "remotes", ["", "remote1,wrong_remote1", "wrong_remote1,wrong_remote2"],
+)
+@pytest.mark.parametrize(
+    "tools", ["", "pidstat,wrong_tool1", "wrong_tool1,wrong_tool2"],
+)
+def test_clear_tools_test86(tools_configuration, groups, remotes, tools):
+    """ Attempt to clear tools at various group-remote-tool locations where the
+    combination may not present on the hosts"""
+    _, tools_tree = tools_configuration
+    command = ["pbench-clear-tools"]
+    if groups:
+        command.append(f"--groups={groups}")
+    if remotes:
+        command.append(
+            f"--remotes={','.join(map('{}.example.com'.format, remotes.split(',')))}"
+        )
+    if tools:
+        command.append(f"--name={tools}")
+
+    out, err, exitcode = pytest.helpers.capture(command)
+    assert b"" == out
+
+    if not groups:
+        groups = "default"
+
+    for group in groups.split(","):
+        err_msg = f'pbench-clear-tools: invalid --group option "{group}"'
+        if group not in tools_tree.keys():
+            assert err_msg.encode() in err
+        else:
+            assert err_msg.encode() not in err
+            for remote in remotes.split(","):
+                err_msg = f'No remote host "{remote}.example.com"'
+                if remote and remote not in tools_tree[group]:
+                    assert err_msg.encode() in err
+                elif remote:
+                    wrong_tools = []
```

Comment:
I don't really like the fact that we're testing `remote` twice.  This really should be structured like:
```
                if remote:
                    if remote not in tools_tree[group]:
                        assert err_msg.encode() in err
                    else:
                        wrong_tools = []
                        ...
                else:
                    remotes = tools_tree[group].keys()
                    ...
```

Although, if you want to "cheat", the following should work:
```
                if remote not in tools_tree[group]:
                    assert err_msg.encode() in err
                elif remote:
                    wrong_tools = []
```
because, if `remote` is an empty string, then the `in` test will return `True`, so it will skip the `if` body and execute the `elif` test, which will send it to the `else` branch.

But, it would probably be best to either take my first suggestion or to completely reorder the cases:
```
                if not remote:
                    remotes = tools_tree[group].keys()
                    ...
                elif remote in tools_tree[group]:
                    wrong_tools = []
                    ...
                else:
                    assert err_msg.encode() in err
```

Reply:
I have updated the code so I guess this should be fixed